p2p/server.py

```diff
@@ -0,0 +1,157 @@
+import asyncio
+import logging
+import secrets
+
+from typing import (
+    List,
+    Tuple,
+)
+
+from eth_keys import datatypes
+
+from eth_utils import big_endian_to_int
+
+from evm.db.chain import AsyncChainDB
+from p2p.auth import (
+    decode_auth_plain,
+    decode_auth_eip8,
+    HandshakeResponder,
+)
+from p2p.cancel_token import CancelToken
+from p2p.constants import (
+    ENCRYPTED_AUTH_MSG_LEN,
+    HASH_LEN,
+)
+from p2p.discovery import DiscoveryProtocol
+from p2p.ecies import ecdh_agree
+from p2p.exceptions import DecryptionError
+from p2p.kademlia import (
+    Address,
+    Node,
+)
+from p2p.peer import (
+    ETHPeer,
+    PeerPool,
+)
+from p2p.utils import sxor
+
+
+class Server:
+    """Server listening for incoming connections"""
+    logger = logging.getLogger("p2p.server.Server")
+    _server = None
+
+    def __init__(self,
+                 privkey: datatypes.PrivateKey,
+                 server_address: Address,
+                 chaindb: AsyncChainDB,
+                 bootstrap_nodes: List[str],
+                 network_id: int
+                 ) -> None:
+        self.cancel_token = CancelToken('Server')
+        self.chaindb = chaindb
+        self.privkey = privkey
+        self.server_address = server_address
+        self.network_id = network_id
+        # TODO: bootstrap_nodes should be looked up by network_id.
+        discovery = DiscoveryProtocol(
+            self.privkey, self.server_address, bootstrap_nodes=bootstrap_nodes)
+        self.peer_pool = PeerPool(ETHPeer, self.chaindb, self.network_id, self.privkey, discovery)
+
+    async def start(self) -> None:
+        self._server = await asyncio.start_server(
+            self.receive_handshake,
+            host=self.server_address.ip,
+            port=self.server_address.tcp_port,
+        )
+
+    async def run(self) -> None:
+        await self.start()
+        self.logger.info("Running server...")
+        await self.cancel_token.wait()
+        await self.stop()
+
+    async def stop(self) -> None:
+        self.logger.info("Closing server...")
+        self.cancel_token.trigger()
+        self._server.close()
+        await self._server.wait_closed()
+        await self.peer_pool.stop()
+
+    async def receive_handshake(
+            self, reader: asyncio.StreamReader, writer: asyncio.StreamWriter) -> None:
+        # Use reader to read the auth_init msg until EOF
+        msg = await reader.read(ENCRYPTED_AUTH_MSG_LEN)
+
+        # Use HandshakeResponder.decode_authentication(auth_init_message) on auth init msg
```

Comment:
This comment reminds that it's a little weird to move `decode_authentication` to `server.py`. To me, it's more organized to leave this kind of functions in `auth.py`.

Reply:
Agreed